pkg/resource/deploy/step.go

```diff
@@ -283,6 +297,36 @@ func (s *DeleteStep) Apply(preview bool) (resource.Status, StepCompleteFunc, err
 	return resource.StatusOK, func() {}, nil
 }
 
+type removePendingReplaceStep struct {
```

Comment:
I'm a bit torn on whether or not this ought to be its own internal step type. We could e.g. make this a special case of `DeleteStep`, which would then obviate the need for the `RemovePendingReplacement` method on `deploy.Events`. However, that would further complicate the code downstream of the engine that consumes these steps, perhaps for little gain.

Thoughts?

Reply:
Hmm... what about if `RemovePendingReplace` was its own step, but it wasn't internal? It seems like it would simply the implementation a little bit:

1. The engine can `Apply` the step as if it were any other step - `Apply` just doesn't do anything
2. There's no need for the `RemovePendingReplacement` event - the snapshot manager can instead react to `RemovePendingReplace` directly without having to involve the rest of the engine
3. The display layer can choose to not ever show `RemovePendingReplace` steps

Does that make sense?

Reply:
> There's no need for the RemovePendingReplacement event - the snapshot manager can instead react to RemovePendingReplace directly without having to involve the rest of the engine

This is what I was getting at above:

> would then obviate the need for the RemovePendingReplacement method

Reply:
Right, but this:

> However, that would further complicate the code downstream of the engine that consumes these steps, perhaps for little gain.

Isn't a problem if it is its own step, right?

Reply:
> The display layer can choose to not ever show RemovePendingReplace step

This is the part that annoys me a bit. This step doesn't really do anything per se; it's really just a signal to the persistence layer to remove the tombstoned resource from the checkpoint (tbh, the fact that we deal with persistence via the `Events` mechanism still bothers me a bit).

Reply:
> Right, but this:
>
>>However, that would further complicate the code downstream of the engine that consumes these steps, perhaps for little gain.
>
> Isn't a problem if it is its own step, right?

That's exactly why it would be bothersome: it would be a new addition to the public Step API, which is relatively broadly exposed.

Reply:
As it turns out, we do already have a mechanism for not reporting steps: I added one so we could elide default provider steps.

Reply:
Yeah... though we already have `Deletes` on `External` resources, which also don't do anything per se except tell the snapshot manager to remove the resource.

I'm fine either way - I do think either rolling this into `Delete` or making it its own step is a little more elegant but I don't feel strongly about it.

Reply:
Either way this feels pretty kludgey. Given that we have some code in the engine to avoid reporting certain steps as events, I feel a little bit better about adding this to the public API for `Step`. I've already written it up.

Reply:
sounds good!